--- /usr/lib/erlang/lib/mnesia-4.3.5/src/mnesia_frag.erl	2007-08-05 08:49:02.000000000 -0700
+++ ./mnesia_frag.erl	2008-06-26 13:44:20.000000000 -0700
@@ -1007,10 +1007,10 @@
 
 remove_node(Node, Cs) ->
     Tab = Cs#cstruct.name,
-    case is_top_frag(Tab) of
-	false ->
+    case ?catch_val({Tab, frag_hash}) of
+	{'EXIT', _} ->
 	    {Cs, false};
-	true -> 
+	_ ->
 	    Pool = lookup_prop(Tab, node_pool),
 	    case lists:member(Node, Pool) of
 		true ->
@@ -1095,14 +1095,6 @@
 	    mnesia:abort({no_exists, Tab, frag_properties, frag_hash})
     end.
 
-is_top_frag(Tab) ->
-    case ?catch_val({Tab, frag_hash}) of
-	{'EXIT', _} ->
-	    false;
-	_ -> 
-	    [] == lookup_foreigners(Tab)
-    end.
-
 %% Returns a list of tables
 lookup_foreigners(Tab) ->
     %% First field in HashPat is either frag_hash or frag_state
