--- /usr/lib/erlang/lib/mnesia-4.3.5/src/mnesia_dumper.erl	2007-08-05 08:49:02.000000000 -0700
+++ ./mnesia_dumper.erl	2008-06-26 13:44:17.000000000 -0700
@@ -826,6 +826,16 @@
 	    end
     end;
 
+insert_op(Tid, _, {op, del_table_copies, NStorList, TabDef}, InPlace, InitBy) ->
+    lists:foreach(fun({Node, Storage}) -> 
+			  insert_op(Tid, 
+				    void, 
+				    {op, del_table_copy, Storage, Node, TabDef},
+				    InPlace,
+				    InitBy)
+		  end,
+		  NStorList);
+
 insert_op(Tid, _, {op, add_table_copy, _Storage, _Node, TabDef}, InPlace, InitBy) ->
     %% During prepare commit, the files was created
     %% and the replica was announced
